File: storio-content-resolver-annotations-processor/src/main/java/com/pushtorefresh/storio/contentresolver/annotations/processor/generate/QueryGenerator.java
package com.pushtorefresh.storio.contentresolver.annotations.processor.generate;

import com.pushtorefresh.storio.contentresolver.annotations.processor.introspection.StorIOContentResolverColumnMeta;
import com.pushtorefresh.storio.contentresolver.annotations.processor.introspection.StorIOContentResolverTypeMeta;

import org.jetbrains.annotations.NotNull;

import java.util.Collections;
import java.util.HashMap;
import java.util.Map;

public class QueryGenerator {

    public static final String WHERE_CLAUSE = "where";
    public static final String WHERE_ARGS = "whereArgs";

    @NotNull
    public static Map<String, String> createWhere(
            @NotNull final StorIOContentResolverTypeMeta storIOContentResolverTypeMeta,
            @NotNull final String varName) {
        final StringBuilder whereClause = new StringBuilder();
        final StringBuilder whereArgs = new StringBuilder();

        int i = 0;

        for (final StorIOContentResolverColumnMeta columnMeta : storIOContentResolverTypeMeta.columns.values()) {
            if (columnMeta.storIOColumn.key()) {
                if (i == 0) {
                    whereClause
                            .append(columnMeta.storIOColumn.name())
                            .append(" = ?");

                    whereArgs
                            .append(varName)
                            .append(".")
                            .append(columnMeta.elementName);
                } else {
                    whereClause
                            .append(" AND ")
                            .append(columnMeta.storIOColumn.name())
                            .append(" = ?");

                    whereArgs
                            .append(", ")
                            .append(varName)
                            .append(".")
                            .append(columnMeta.elementName);
                }

                if (columnMeta.isMethod()) {
                    whereArgs.append("()");
                }

                i++;
            }
        }

        if (whereClause.length() == 0 || whereArgs.length() == 0) {
            return Collections.emptyMap();
        } else {
            final Map<String, String> result = new HashMap<String, String>(2);

            result.put(WHERE_CLAUSE, whereClause.toString()); // example: "email = ? AND user_id = ?"
            result.put(WHERE_ARGS, whereArgs.toString()); // example: "object.email, object.userId"

            return result;
        }
    }
}

File: storio-content-resolver-annotations-processor/src/main/kotlin/com/pushtorefresh/storio/contentresolver/annotations/processor/generate/QueryGenerator.kt
package com.pushtorefresh.storio.contentresolver.annotations.processor.generate

import com.pushtorefresh.storio.contentresolver.annotations.processor.introspection.StorIOContentResolverTypeMeta

object QueryGenerator {

    val WHERE_CLAUSE = "where"
    val WHERE_ARGS = "whereArgs"

    fun createWhere(typeMeta: StorIOContentResolverTypeMeta, varName: String): Map<String, String> {
        val whereClause = StringBuilder()
        val whereArgs = StringBuilder()

        var i = 0

        typeMeta.columns.values.forEach { columnMeta ->
            if (columnMeta.storIOColumn.key) {
                if (i == 0) {
                    whereClause
                            .append(columnMeta.storIOColumn.name)
                            .append(" = ?")

                    whereArgs
                            .append(varName)
                            .append(".")
                            .append(columnMeta.contextAwareName)
                } else {
                    whereClause
                            .append(" AND ")
                            .append(columnMeta.storIOColumn.name)
                            .append(" = ?")

                    whereArgs
                            .append(", ")
                            .append(varName)
                            .append(".")
                            .append(columnMeta.contextAwareName)
                }
                i++
            }
        }

        return when {
            whereClause.isEmpty() || whereArgs.isEmpty() -> emptyMap()
            else -> mapOf(
                    WHERE_CLAUSE to whereClause.toString(),
                    WHERE_ARGS to whereArgs.toString())
        }
    }
}